File: android/core/data/src/main/kotlin/com/riskdetectedan/core/data/notebook/NotebookOrganization.kt
package com.riskdetectedan.core.data.notebook

import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable data class NotebookItem(val item_id: String, val text: String, val done: Boolean)
@Serializable data class NotebookOrganization(val schema_version: Int, val note_id: String, val version: Long,
    val tombstone: Boolean, val items: List<NotebookItem>, val tags: List<String>) {
    fun validate(id: String) {
        require(schema_version == 1 && note_id == id && version in 1..9007199254740991L && (!tombstone || (items.isEmpty() && tags.isEmpty())))
        validateNotebookOrganization(items, tags)
    }
}
fun validateNotebookOrganization(items: List<NotebookItem>, tags: List<String>) {
    require(items.size <= 500 && tags.size <= 30 && items.map { it.item_id }.toSet().size == items.size)
    require(items.all { UUID.fromString(it.item_id).toString() == it.item_id && it.text.isNotBlank() && it.text.codePointCount(0, it.text.length) <= 1000 })
    require(tags.all { it.isNotBlank() && it.codePointCount(0, it.length) <= 60 })
    require(tags.map { it.trim().lowercase(java.util.Locale.ROOT) }.toSet().size == tags.size)
}

File: App/Services/Notebook/NotebookOrganization.swift
import Foundation

struct NotebookItem: Codable, Equatable {
    let item_id: UUID
    var text: String
    var done: Bool
}
struct NotebookOrganization: Codable {
    let schema_version: Int; let note_id: UUID; let version: Int64; let tombstone: Bool
    let items: [NotebookItem]; let tags: [String]
    func validate(_ id: UUID) throws {
        guard schema_version == 1, note_id == id, (1...9007199254740991).contains(version),
              !tombstone || (items.isEmpty && tags.isEmpty) else { throw NotebookFailure.invalid }
        try validateNotebookOrganization(items, tags)
    }
}
func validateNotebookOrganization(_ items: [NotebookItem], _ tags: [String]) throws {
    guard items.count <= 500, tags.count <= 30, Set(items.map(\.item_id)).count == items.count,
          items.allSatisfy({ !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && $0.text.unicodeScalars.count <= 1000 }),
          tags.allSatisfy({ !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && $0.unicodeScalars.count <= 60 }),
          Set(tags.map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }).count == tags.count else { throw NotebookFailure.invalid }
}
